website/www/site/content/en/documentation/patterns/state-and-timers.md

```diff
@@ -0,0 +1,143 @@
+
+The result in Python:
+
+```python
+class _BatchItems(DoFn):
+    elements_bag = BagStateSpec('buffer', IterableCoder(PickleCoder()))
+    event_timer = TimerSpec('event_timer', TimeDomain.WATERMARK)
+
+    def __init__(self, max_batch_size=5000):
+        self.max_batch_size = max_batch_size
+
+    def process(self,
+                element,
+                window=DoFn.WindowParam,
+                buffer_state=DoFn.StateParam(elements_bag),
+                event_timer=DoFn.TimerParam(event_timer)):
+
+        buffer_state.add(element)
+        event_timer.set(window.max_timestamp())
+
+    @on_timer(event_timer)
+    def expiry(self, buffer_state=DoFn.StateParam(elements_bag), event_timer=DoFn.TimerParam(event_timer)):
+        batch = [element for element in buffer_state.read()]
+        if not batch:
+            return
+
+        key, _ = batch[0]
+        current_batch_size = itertools.count()
+        output = []
+
+        if batch:
+            for event in batch:
+                clear_buffer = next(current_batch_size) >= self.max_batch_size
+                if clear_buffer:
+                    yield (key, output)
+                    output.clear()
+                    current_batch_size = itertools.count()
+                output.append(event[1])
+
+        if len(output) > 0:
+            yield (key, output)
+```
+
+and in Java:
+
+```java
+class BatchRequestForRecommendationAI extends DoFn<KV<String, GenericJson>, KV<String, Iterable<GenericJson>>> {
+    private static final Logger LOG = LoggerFactory.getLogger(BatchRequestForRecommendationAI.class);
+
+    private final Counter numberOfRowsBagged =
+            Metrics.counter(BatchRequestForRecommendationAI.class, "numberOfRowsBagged");
+
+    private final Integer maxBatchSize;
+
+    @StateId("elementsBag")
+    private final StateSpec<BagState<KV<String, GenericJson>>> elementsBag = StateSpecs.bag();
+
+    @TimerId("eventTimer")
+    private final TimerSpec eventTimer = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+    public BatchRequestForRecommendationAI(Integer maxBatchSize) {
+        this.maxBatchSize = maxBatchSize;
+    }
+
+    @ProcessElement
+    public void process(
+            @Element KV<String, GenericJson> element,
+            @StateId("elementsBag") BagState<KV<String, GenericJson>> elementsBag,
+            @TimerId("eventTimer") Timer eventTimer,
+            BoundedWindow w) {
+        elementsBag.add(element);
+        eventTimer.set(w.maxTimestamp());
+    }
+
+    @OnTimer("eventTimer")
+    public void onTimer(
+            @StateId("elementsBag") BagState<KV<String, GenericJson>> elementsBag,
+            OutputReceiver<KV<String, Iterable<GenericJson>>> output) {
+        if (elementsBag.read().iterator().hasNext()) {
+            String key = elementsBag.read().iterator().next().getKey();
+            AtomicInteger currentBatchSize = new AtomicInteger();
```

Comment:
Why use AtomicInteger here?

Reply:
In case it is accessed by multiple threads, but since we're scoped to one key I guess that's not necessary?

Reply:
Yes, would only be needed if you had a static object which all the DoFn's on a machine had access to.